plugins/inputs/dpdk/dpdk.go

```diff
@@ -0,0 +1,240 @@
+		if len(commandWithParams) >= maxCommandLengthWithParams {
+			return fmt.Errorf("command with parameters '%v' shall be less than %v characters", commandWithParams, maxCommandLengthWithParams)
+		}
+	}
+
+	return nil
+}
+
+// Gathers all unique commands and processes each command sequentially
+// Parallel processing could be achieved by running several instances of this plugin with different settings
+func (dpdk *dpdk) Gather(accumulator telegraf.Accumulator) error {
+	commands := dpdk.gatherCommands(accumulator)
+
+	for _, command := range commands {
+		dpdk.processCommand(command, accumulator)
+	}
+
+	return nil
+}
+
+// Gathers all unique commands
+func (dpdk *dpdk) gatherCommands(accumulator telegraf.Accumulator) []string {
+	var commands []string
+	if contains(dpdk.DeviceTypes, "ethdev") {
+		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.ethdevCommands, dpdk.EthdevConfig.EthdevExcludeCommands, ethdevListCommand, accumulator)...)
+	}
+
+	if contains(dpdk.DeviceTypes, "rawdev") {
+		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.rawdevCommands, []string{}, rawdevListCommand, accumulator)...)
+	}
+
+	commands = append(commands, dpdk.AdditionalCommands...)
+
+	return uniqueValues(commands)
+}
+
+func (dpdk *dpdk) getCommandsAndParamsCombinations(appendedCommands []string, excludedCommands []string, listCommand string, accumulator telegraf.Accumulator) []string {
+	validCommands := removeSubset(appendedCommands, excludedCommands)
+
+	commands, err := dpdk.appendCommandsWithParamsFromList(listCommand, validCommands)
+	if err != nil {
+		accumulator.AddError(fmt.Errorf("error occurred during fetching of %v params - %v", listCommand, err))
+	}
+	return commands
+}
+
+// Fetches all identifiers of devices and then creates all possible combinations of commands for each device
+func (dpdk *dpdk) appendCommandsWithParamsFromList(listCommand string, commands []string) ([]string, error) {
+	response, err := dpdk.connector.getCommandResponse(listCommand)
+	if err != nil {
+		return nil, err
+	}
+
+	params, err := jsonToArray(response, listCommand)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]string, 0, len(commands)*len(params))
+	for _, command := range commands {
+		for _, param := range params {
+			result = append(result, commandWithParams(command, param))
+		}
+	}
+
+	return uniqueValues(result), nil
+}
+
+// Executes command, parses response and creates/writes metric from response
+func (dpdk *dpdk) processCommand(commandWithParams string, accumulator telegraf.Accumulator) {
+	buf, err := dpdk.connector.getCommandResponse(commandWithParams)
+	if err != nil {
+		accumulator.AddError(err)
+		return
+	}
+
+	var parsedResponse map[string]interface{}
+	err = json.Unmarshal(buf, &parsedResponse)
+	if err != nil {
+		accumulator.AddError(fmt.Errorf("failed to unmarshall json response from %v command - %v", commandWithParams, err))
+		return
+	}
+
+	command := stripParams(commandWithParams)
```

Comment:
Hmm wouldn't it make more sense to concatenate `command` and `params` when calling `getCommandResponse` instead of first fuse the two and then split them here again? You could even prepare the params to be either `""` or `",param1..."` and simply do `buf, err := dpdk.connector.getCommandResponse(command+preparedParams)`...

Reply:
This was done intentionally to simplify code, because before execution of commands we have to leave only unique values. This could be done with approach you mentioned (separated command and params), but it will be much more complicated code than it is now, especially since splitting operation is pretty simple (both in code and in algorithmic complexity).

I wouldn't change it.

Reply:
Hmmm ok but I don't see the problem in making the stuff unique. If you really insist, keep it as is... :-)

Reply:
Thank you, keeping it as is :)